sys/ztimer/util.c

```diff
@@ -180,3 +181,18 @@ int ztimer_mutex_lock_timeout(ztimer_clock_t *clock, mutex_t *mutex,
     ztimer_remove(clock, &t);
     return 0;
 }
+
+int ztimer_rmutex_lock_timeout(ztimer_clock_t *clock, rmutex_t *rmutex,
+                              uint32_t timeout)
+{
+    if (rmutex_trylock(rmutex)) {
+        return 0;
+    }
+    if (ztimer_mutex_lock_timeout(clock, &rmutex->mutex, timeout) == 0) {
+        atomic_store_explicit(&rmutex->owner,
+                              thread_getpid(), memory_order_relaxed);
+        rmutex->refcount++;
```

Comment:
The refcount needs also to be incremented atomically to prevent corruptions. E.g. incrementing on an 8bit platform to a value of `0x00ff` will contain after the first CPU cycle a value of `0x0000` and only after a second cycle `0x0100`.

But since this issue is consistent with the code upstream, this should be fixed elsewhere first. Maybe we should drop the use of C11 atomics for just using `irq_disable()` ... `irq_restore()`. (I think @kaspar030 already suggested doing this on a different occasion. I guess, I should listen to him more :-))

Reply:
I'm not sure I understand how there can be corruption on the refcount as it is modified only by the thread that has the mutex. Am I missing something?

Reply:
Aha, yes, I overlooked the mutex :-) I wonder why there both mutex and C11 atomics in use. Maybe it would be indeed good to take a second look at the implementation.

Reply:
OK, it makes sense to do this atomically, as when locking the mutex inside the rmutex fails, this could be due to the caller already holding the rmutex. Hence, we need to verify the owner of the mutex, but unprotected by the mutex.